📖 Runtime SDK proposal (#6181)
kubernetes-sigs/cluster-api: docs/proposals/20220221-runtime-SDK.md

    - [Discovering Runtime Extensions](#discovering-runtime-extensions)
    - [Calling Runtime Extensions](#calling-runtime-extensions)
  - [Security Model](#security-model)
  - [Risks and Mitigations](#risks-and-mitigations)
  - [Alternatives](#alternatives)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
    - [Test Plan](#test-plan)
    - [Graduation Criteria](#graduation-criteria)
    - [Version Skew Strategy](#version-skew-strategy)
  - [Annex](#annex)
    - [Runtime SDK rules](#runtime-sdk-rules-1)
  - [Implementation History](#implementation-history)
  
## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

- **Cluster API Runtime**: identifies the Cluster API execution model, a set of controllers cooperating in managing the 
  Workload Cluster’s lifecycle.
- **Runtime SDK**: a set of rules, recommendations and fundamental capabilities required to develop Runtime Hooks and 
  Runtime Extensions.
- **Runtime Hook**: a single, well identified, extension point allowing applications built on top of Cluster API to hook

>   into specific moments of the Workload Cluster’s lifecycle, e.g. Cluster.BeforeUpgrade.

Reviewer: >into specific moments of the Workload Cluster’s lifecycle

Does this include both Cluster and Machine CRs or something different?

Author: This includes everything in a Cluster, including the control plane and worker machines, but eventually, also MHC and other resources are relevant for the Cluster lifecycle

Author: Added an example to clarify this

### Non-Goals

- To identify or specify the list of Runtime Hooks that should be implemented; some examples of possible Runtime Hooks
  will be eventually provided, but it is not in the scope of this document to define them in detail;
- To replace controllers or any other component of the Cluster API Runtime (including infrastructure providers,
  bootstrap providers, control plane providers and the CRD/contract based extension mechanism they rely on).
- To support the implementation of Runtime Hooks anywhere else outside the main Cluster API repository.

### Future Work

- Identify and specify the list of Runtime Hooks to be implemented; this will be addressed iteratively by a set of
  future proposals, all of them building on top of the foundational capabilities introduced by this document;
- Eventually consider deprecation of machine deletion hooks and replacement with a Runtime Hook;
- Improve the Runtime Extension developer guide based on experience and feedback;
- Add metrics about Runtime Extension calls (usage, usage vs deprecated versions, duration, error rate etc.);
- Allow providers to use the same SDK to define their own hooks.

## Proposal

### User Stories

> - As a cluster operator I want to be able to execute a particular action in well-defined moments of the Workload

Reviewer: Can we give a "particular action" concrete example so is immediately clear?

Reviewer 2: +1 on concrete examples (even multiple would be nice)

Author: added a couple of examples. we can eventually make this list grow.

Author: @enxebre @CecileRobertMichon 
added a couple of examples; while doing this I also added a use case for externally generated patches, which is another potential use case for this SDK even if it does not qualify as a lifecycle event; I have tried to qualify this in a couple of point of the proposal (lifecycle hooks are the main goal, other usage are _by opportunity_)

####  Alternative deployments methods
Alternative deployment methods can be used given that the requirement about the HTTP endpoint accessibility
is satisfied, like e.g.

- deploying the HTTP Server as a part of another component, e.g. a controller;
- deploying the HTTP Server outside of the Management Cluster.

- Also in this case above recommendations about availability and identity and access management apply.

### Registering Runtime Extensions

_Important! Cluster administrators should carefully vet any Runtime Extension registration, thus preventing
malicious components from being added to the system._

_Creating RuntimeExtensionConfiguration will be allowed only if the Runtime Extension feature flag is set to true._

By registering a Runtime Extension the Cluster API Runtime became aware of a Runtime Extension implementing a
Runtime Hook, and as a consequence the runtime starts calling the extension at well-defined moments of the
Workload Cluster’s lifecycle.

This process has many similarities with registering dynamic webhooks in Kubernetes, but in this case it is required
to use the following CR:

> 

Reviewer: Is this an example with pseudo API for reference or this proposal including a new RuntimeExtensionConfiguration API kind as it looks below?

Author: This proposal includes a new RuntimeExtensionConfiguration kind as it looks below +/- improvements if we get agreement

  timeoutSeconds: 2
  # FailurePolicy defines how errors from RuntimeExtension calls are handled - allowed values are Ignore or Fail. Defaults to Fail.
  failurePolicy: Fail
```

Following elements can be added to the above type if consensus is reached/in a follow up iteration:

- Support for defining Runtime Extensions that applies to a subset of Clusters only, e.g.
```yaml
# NamespaceSelector decides whether to run the webhook on a Cluster based on whether the namespace for that Cluster matches the selector.
# If not specified, the WebHook runs for all the namespaces.
namespaceSelector:
...

# Or

# ClusterSelector decides whether to run the webhook on a Cluster based on whether the Cluster object matches the selector.
# If not specified, the WebHook runs for all the Clusters.
clusterSelector:
...

# Or a combination of both criteria combined in AND.
```
- Support for defining dependencies among Runtime Extensions, e.g. taking inspiration from [systemd unit options](https://www.freedesktop.org/software/systemd/man/systemd.unit.html)
  Note: designing Runtime Extensions without dependency should always be preferred.

> 

Reviewer: I'd be hesitant to introduce complexity with a dependency system and introduce "pet" like state, unless there's a strong and evident need for it which can be solved with different angle. May be we can consider dropping this for now to avoid confusion?

Author: Agreed. created a section with what we don't want to add at this stage/with your considerations, however, I would like to keep it around for tracking that the idea has been discussed

    - [Discovering Runtime Extensions](#discovering-runtime-extensions)
    - [Calling Runtime Extensions](#calling-runtime-extensions)
  - [Security Model](#security-model)
  - [Risks and Mitigations](#risks-and-mitigations)
  - [Alternatives](#alternatives)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
    - [Test Plan](#test-plan)
    - [Graduation Criteria](#graduation-criteria)
    - [Version Skew Strategy](#version-skew-strategy)
  - [Annex](#annex)
    - [Runtime SDK rules](#runtime-sdk-rules-1)
  - [Implementation History](#implementation-history)
  
## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

- **Cluster API Runtime**: identifies the Cluster API execution model, a set of controllers cooperating in managing the 
  Workload Cluster’s lifecycle.
- **Runtime SDK**: a set of rules, recommendations and fundamental capabilities required to develop Runtime Hooks and 
  Runtime Extensions.
- **Runtime Hook**: a single, well identified, extension point allowing applications built on top of Cluster API to hook

>   into specific moments of the Workload Cluster’s lifecycle, e.g. Cluster.BeforeUpgrade.

Author: Given that this proposal is introducing a lot of new concepts and primitives, I have tried to keep definitions as simple as possible while documenting nuances/variants down in the doc.

WRT to this specific point, at the current stage, most of the use cases we are considering are about lifecycle hooks while instead, we have only one very specific use case for extending reconcile, so I have focused the definition on Lifecycle Hooks only and documented reconcile hooks down in the doc as something that we are going to do by opportunity.
 
However, if someone else feels that documenting runtime hooks variant in the definition is relevant and doesn't create confusion for the users at this stage of the document, I'm happy to accept the suggestion

>   into specific moments of the Workload Cluster’s lifecycle, e.g. Cluster.BeforeUpgrade.

Reviewer 3: Sounds good to me!

- **Runtime Extension**: an external component which is part of a system/product built on top of Cluster API that can 
  handle requests for a specific Runtime Hook.

## Summary

This proposal introduces the Cluster API Runtime SDK, a set of rules, recommendations and fundamental capabilities
required to implement a new extensibility mechanism that will allow systems, products and services built on top of

> Cluster API to interact with the Workload Cluster’s lifecycle.

Author: Fixed
WRT to documenting variants at this stage of the document, same considerations as above applies

- To define the rules ensuring Runtime Hooks can evolve over time:
  - When/how to modify the current version;
  - When/how to create a new version;
  - When/how to deprecate an old version, as well as mechanisms to inform users about versions being deprecated;
  - When/how to drop an old version, as well as providing a mechanism to prevent users to upgrade Cluster API when
    this will break installed Runtime Extensions;
- To define the fundamental capabilities/tooling to be implemented in CAPI in order to allow the implementation of
  Runtime Hooks.
- To provide an initial set of guidelines for Runtime Extension developers.
- To define how external Runtime Extensions can be registered within the Cluster API Runtime.

### Non-Goals

- To identify or specify the list of Runtime Hooks that should be implemented; some examples of possible Runtime Hooks
  will be eventually provided, but it is not in the scope of this document to define them in detail;
- To replace controllers or any other component of the Cluster API Runtime (including infrastructure providers,
  bootstrap providers, control plane providers and the CRD/contract based extension mechanism they rely on).

> - To support the implementation of Runtime Hooks anywhere else outside the main Cluster API repository.

Author: done

### Future Work

- Identify and specify the list of Runtime Hooks to be implemented; this will be addressed iteratively by a set of
  future proposals, all of them building on top of the foundational capabilities introduced by this document;
- Eventually consider deprecation of machine deletion hooks and replacement with a Runtime Hook;
- Improve the Runtime Extension developer guide based on experience and feedback;
- Add metrics about Runtime Extension calls (usage, usage vs deprecated versions, duration, error rate etc.);
- Allow providers to use the same SDK to define their own hooks.

## Proposal

### User Stories

- As a cluster operator I want to be able to execute a particular action in well-defined moments of the Workload

>   Cluster’s lifecycle, e.g. 

Author: I have considered breaking the last point into a separate user story but then dropped the idea because reconciliation at the end is an implementation detail from the user PoV "computing the Cluster topology" could be considered a Cluster Lifecycle event that happens continuously.

e.g. in the example above the `Cluster.BeforeUpgrade` operations exist in version `v1alpha1` (old version)
and `v1alpha2` (current).

Supporting more versions at the same time is a requirement in order to:

- Allow Cluster API maintainers to continue to develop and evolve Runtime Hooks in a predictable way.
- Provide a well-defined set of guarantees to Runtime Extension implementers they can rely on while developing
  solutions on top of CAPI.

In their simplest form guarantees for Runtime Hooks versions are:

- Once a Runtime Hook version has been published, breaking changes are not allowed without bumping to a new version
  (e.g. fields removal/renaming)
- Before removing a Runtime Hook version, a deprecation period should be respected, with a duration depending
  on the maturity of the API itself (12 Months/3 Versions GA, 6M/2V beta, 0 alpha).

The formal definition of the Runtime SDK rules derived from https://kubernetes.io/docs/reference/using-api/deprecation-policy/,
can be found in the annex at the end of the document. Please note that during implementation we will consider a
mechanism allowing to:

> - Inform admins about Runtime Extension using a deprecated version of a Runtime Hook (e.g. return a well known

Author: Added as an option.
However, webhook warnings apply does not alert you about existing extensions being deprecated unless you change the API object

> - Inform admins about Runtime Extension using a deprecated version of a Runtime Hook (e.g. return a well known

Reviewer 3: That's correct. For that it would be good to have warnings in clusterctl or maybe even on Provider CR create/update.

I'm just not sure a HTTP header somewhere would be noticed. But I'm just not really aware if there are well-known headers for something like that.

  of choice, and we are not planning to test/provide tooling/libraries for other languages. Nevertheless, given that
  we rely on Open API and plain HTTP(s) calls, other languages should just work but support will be provided at
  best effort.
- In case it is required to implement more than one extension, to choose if to have a single HTTP server with multiple
  handlers or separated HTTP servers;
- To choose if to have a dedicated HTTP server(s) for Runtime Extensions only or if to use the HTTP server for other
  purposes as well (e.g. metric endpoint).

In case the Runtime Extension is being developed in Golang, the implementer can benefit from importing
`sigs.k8s.io/cluster-api` and:

- Use Golang types defined under `/runtime` (the types from which the OpenAPI specification has been generated).
- Use the RuntimeExtension catalog object to generate the skeleton of the HTTP handler for a given Runtime extension.
  The generated func will take care of scaffolding tasks like Marshal/Unmarshal; the only missing part will be to
  implement a strongly typed func that contains the actual extension implementation.

E.g.

> _Note: this is pseudo code, meant to demonstrate that implementing a RuntimeExtension requires only minimal scaffolding;

Reviewer 3: nit: Markdown renders those two lines in one

> E.g. Note: this is pseudo code, meant to demonstrate that implementing a RuntimeExtension requires only minimal scaffolding; the exact details will be defined during implementation, possibly taking advantage of Golang generics.

(just in case that's not intended)

> the exact details will be defined during implementation, possibly taking advantage of Golang generics._

Author: Agreed, we can figure this out during implementation

    // your actual implementation...
    return nil
}
```

While writing the actual code of the Runtime Extension a set of important guidelines must be considered:

#### Timeouts
Runtime Extension processing adds to network request latency, they should run as quickly as possible
(typically in milliseconds); Cluster Administrator will be allowed to configure how long the Cluster API Runtime
should wait for a Runtime Extension to respond before treating the call as a failure (max 10s).

####  Availability
Runtime Extension failure could result in errors in handling Workload’s Clusters lifecycle, and so the implementation
should be robust, have proper error handling, avoid panics, etc.; Cluster Administrator will be allowed to set up
failure policies preventing a Runtime Extension failure to have negative effects on the Cluster API Runtime, but
this option can’t be used in all the use cases.

#### Side Effects
It is recommended that Runtime Extensions should avoid side effects if possible, which means to operate only on
the content of the Input/Output sent to them, and not make out-of-band changes.

> If side effects are required, following rules apply:

Author: Fixed

#### Idempotence
An idempotent Runtime Extension is able to successfully accomplish its task also in case it has already been completed
(the Runtime Extension checks current state and changes it only if necessary).

#### Avoid dependencies
Each Runtime Extension should accomplish its task without dependency or relations with other Runtime Extensions.
Introducing dependencies across Runtime Extensions makes the system fragile, and it is probably a consequence of
poor “Separation of Concerns” while designing such components.

> 

Reviewer 3: Q: Should we mention determinism here too?

I think we don't have to, but for some RuntimeExtensions it might be super important that the same input always results in the same output.

Reviewer 3: Sounds good!

#### Availability
It is recommended that Runtime Extensions should leverage some form of load-balancing, to provide high availability
and performance benefits; you can run multiple Runtime Extension backends behind a service to leverage the
load-balancing that services support.

#### Identity and access management
The security model for each Runtime Extensions should be carefully defined, like for any other application deployed
in the Cluster: the deployment must use a dedicated service account with limited RBAC permission.

On top of that, the container image for the Runtime Extension should be carefully designed in order to avoid
privilege escalation (e.g using [distroless](https://github.com/GoogleContainerTools/distroless) base images) and
the Pod spec in the Deployment manifest should enforce security best practices (e.g. do not use privileged pods etc.).

####  Alternative deployments methods
Alternative deployment methods can be used given that the requirement about the HTTP endpoint accessibility
is satisfied, like e.g.

- deploying the HTTP Server as a part of another component, e.g. a controller;
- deploying the HTTP Server outside of the Management Cluster.

> - Also in this case above recommendations about availability and identity and access management apply.

Author: done

### Registering Runtime Extensions

_Important! Cluster administrators should carefully vet any Runtime Extension registration, thus preventing
malicious components from being added to the system._

_Creating RuntimeExtensionConfiguration will be allowed only if the Runtime Extension feature flag is set to true._

By registering a Runtime Extension the Cluster API Runtime became aware of a Runtime Extension implementing a
Runtime Hook, and as a consequence the runtime starts calling the extension at well-defined moments of the
Workload Cluster’s lifecycle.

This process has many similarities with registering dynamic webhooks in Kubernetes, but in this case it is required
to use the following CR kind introduced by this Proposal:

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: RuntimeExtensionConfiguration
metadata:

> name: "my-amazing-product-runtime-extension"

Reviewer 3: ```suggestion
  name: "my-amazing-product-runtime-extension"
```
nit

webhooks:
# Name should be fully qualified and unique in the Cluster, thus usage of sub domains, version qualifiers is recommended.
- name: "my-amazing-runtime-extension.v1.5.panda.com"
  # List of group/version/hook the RuntimeExtension implements. Required
  operations:

>   - apiVersion: "cluster.runtime.cluster.x-k8s.io/v1alpha2"

Author: agreed, we should figure this out when we start discussing hook implementation; for the SDK this is an invariant

[systemd unit options](https://www.freedesktop.org/software/systemd/man/systemd.unit.html) or alternative approaches.

The main reason behind that is that such type of feature introduces complexity and creates "pet" like relations across
components making the overall system more fragile. This is also consistent with the [avoid dependencies](#avoid-dependencies) 
recommendation above.

## Runtime Hooks developer guide (CAPI internals)

_Following notes provide details about how Runtime Hook will be implemented in the Cluster API codebase;
if you are not interested in CAPI internals you can skip this section._

### Runtime hook implementation

The process of implementing the new Runtime Hooks is intentionally designed in order to mimic the steps currently
used to define API types, thus providing a familiar experience to the maintainers/the people used to look at the
Cluster API codebase. Most specifically:

- Runtime Hooks versions must be defined under a `/runtime` folder.
- For each Runtime Hook, there must be one version, each one defined in its own folder e.g. `/v1alpha1`, `/v1alpha2` etc.

```
/runtime
└── contract

>     ├── cluster

Author: Clarified this is an option we can consider later on

responsibility of this controller should be to maintain an internal, shared **registry** of available extensions
at a given time.

Please note that the Runtime Extensions registry also provides a single point to centralize a set of common behaviors
supporting interaction with those external components, thus making the adoption of this feature scalable - 
in the sense of being used for an increasing numbers of use cases in Cluster API - while operating consistently
across the board.

A first behavior that falls into this category is the implementation of exponential backoff mechanisms
in case of errors, thus preventing Cluster API from creating pressure on HTTP Servers recovering from or with
ongoing operational issues.

Another cross-cutting concern is about ensuring that Runtime Extensions, which are external components triggered
in the middle of Cluster API controllers logic, do not block the reconciliation process indefinitely 
(e.g by enforcing a maximum timeout for all the Runtime Extensions calls).

### Calling Runtime Extensions

_Note: the code described in this paragraph will be executed only if the Runtime Extension feature flag is set to true._

> Cluster API is going to implement calls to registered Runtime Extensions at well-known moments of the Cluster’s lifecycle.

Reviewer 3: ```suggestion
Cluster API is going to implement calls to registered Runtime Extensions at well-known moments.
```
Maybe we can just simplify a bit here, and by that keep the door a bit wider open to non-lifecycle-related hooks

Author: Same consideration as above, let's focus on the main use case for the sake of simplicity

// the exact details will be defined during implementation, possibly taking advantage of Golang generics.

var c = catalog.NewCatalog()

func init() {
    v1alpha2.AddToCatalog(c)
}

func main() {
    ctx := context.Background()

    listener, err := net.Listen("tcp", net.JoinHostPort("127.0.0.1", "8082"))
    if err != nil {
        panic(err)
    }

    // Build an HTTP handler for a given operation, calling a strongly typed func at each request
    beforeUgradeHandler, err := catalogHTTP.NewHandlerBuilder().
        WithCatalog(c).
        AddService(&v1alpha2.BeforeUgradeHook{}, doBeforeUpgrade).
        Build()
    if err != nil {
        panic(err)
    }

    srv := &http.Server{
        Handler: BeforeUgradeHandler,
    }
    if err := srv.Serve(listener); err != nil && err != http.ErrServerClosed {
        panic(err)
    }
}

func doBeforeUpgrade(in *v1alpha2.BeforeUpgradeInput, out *v1alpha2.BeforeUpgradeOutput) error {
    // your actual implementation...
    return nil
}
```

While writing the actual code of the Runtime Extension a set of important guidelines must be considered:

> 

Reviewer: should we have an "errors" section? what happens when one or more extensions fail? how do we react to the different http response? Do we need a .failurePolicy in the API?

Reviewer: I just saw .FailurePolicy is include already, still probably worth it an error section here.

Author: Is it ok if I iterate on this while we figure out how to handle HTTP return codes (we are investigating how this is implemented in the API server for dynamic admission webhooks)

Reviewer 3: Yup would be good if we can iterate a bit on it. Overall we're trying to make everything more concrete over time when we figure out how to actually use the Runtime SDK to implement external patches / lifecycle hooks.

Reviewer 2: IMO what happens when things fail is a critical part of the design and we shouldn't leave it for later. Seems like this proposal already is introducing opinions on the topic (failurePolicy, timeout) so it would make sense to summarize in an error section as @enxebre pointed out. We can list the stuff that is still TBD / future work in that section as well.

Author: @CecileRobertMichon @enxebre 
I have documented error management as well as how blocking runtime extensions should work

- name: "my-amazing-runtime-extension.v1.5.panda.com"
  # List of group/version/hook the RuntimeExtension implements. Required
  operations:
  - apiVersion: "cluster.runtime.cluster.x-k8s.io/v1alpha2"
    hook: "beforeUpgrade"
  - apiVersion: "cluster.runtime.cluster.x-k8s.io/v1alpha2"
    hook: "afterUpgrade"
  #ClientConfig defines how to communicate with the RuntimeExtension. Required
  clientConfig:
    #`url` gives the location of the RuntimeExtension, in standard URL form (`scheme://host:port/path`). Exactly one of `url` or `service` must be specified.
    url: "..."
    service:
      namespace: "example-namespace"
      name: "example-service"
      # `path` is an optional path prefix path which can be sent in any request to this service.
      path: "runtime-extensions/"
      # If specified, the port on the service that hosts the RuntimeExtension. Default to 443. `port` should be a valid port number (1-65535, inclusive).
      port: 8082
    caBundle: "..."
  # If specified, define the timeout for each RuntimeExtension call to complete (Default 5s, Max 10s).
  timeoutSeconds: 2

>   # FailurePolicy defines how errors from RuntimeExtension calls are handled - allowed values are Ignore or Fail. Defaults to Fail.

Reviewer: may be convert this yaml to a struct definition so the API concrete types can be discussed, thought I think the idea is clear and will get there iteratively as we start implementing it, so I'm fine either way.

Reviewer 3: I'm not sure it's helpful if we try to discuss the Go structs here, given that we usually do it twice. And to be honest we take a far closer look in the implementation PRs I saw so far.

Author: I'm ok with both solutions.
What if we link the types once they merge?

Reviewer 3: Sounds like a good and maintainable approach going forward.

>   failurePolicy: Fail

Reviewer: should the failure policy be at the group/version/hook level?

Author: @enxebre that's a good point.
while investigating this a little bit more, we came to the conclusion that it is convenient to have a RuntimeExtension server to server multiple runtime extensions, and this quickly scales to 10-20 extensions; in order to avoid putting on the cluster administrator the burden to document/maintain such a list, we decided to propose a discovery mechanism, thus shifting this responsibility to the extensions developers. 
I have updated the document accordingly, PTAL

      - [Deterministic result](#deterministic-result)
    - [Deploy Runtime Extensions](#deploy-runtime-extensions)
      - [Availability](#availability-1)
      - [Identity and access management](#identity-and-access-management)
      - [Alternative deployments methods](#alternative-deployments-methods)
    - [Registering Runtime Extensions](#registering-runtime-extensions)
  - [Runtime Hooks developer guide (CAPI internals)](#runtime-hooks-developer-guide-capi-internals)
    - [Runtime hook implementation](#runtime-hook-implementation)
    - [Discovering Runtime Extensions](#discovering-runtime-extensions)
    - [Calling Runtime Extensions](#calling-runtime-extensions)
  - [Security Model](#security-model)
  - [Risks and Mitigations](#risks-and-mitigations)
  - [Alternatives](#alternatives)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
    - [Test Plan](#test-plan)
    - [Graduation Criteria](#graduation-criteria)
    - [Version Skew Strategy](#version-skew-strategy)
  - [Annex](#annex)
    - [Runtime SDK rules](#runtime-sdk-rules-1)
  - [Implementation History](#implementation-history)
  
## Glossary

> Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

Reviewer 3: Q: At which point are we moving new definitions into the glossary and the book? (and don't mention them explicitly anymore in follow-up proposals)

Author: Good question.
I think we have to make a one-off pass into all the old proposals, and then make upgrading the glossary part of the proposal work (might be point this out in the proposal template)

A nice side effect of this choice is the possibility to leverage on a set api-machinery tooling and practices the
Cluster API maintainers are well-used to.

It is also important to notice that the model based on Runtime Hooks and Runtime Extensions implies two separate
personas being involved, each one with its own responsibilities in the process:

![overview](images/runtime-sdk/overview.png)

The Runtime SDK rules defined in this document are a critical element of the above split of responsibilities,
defining expectations for each of the above personas.

#### Cluster API Runtime Hooks vs Kubernetes admission webhooks

Runtime Hooks are inspired by Kubernetes admission webhooks, but there is one key difference that splits them apart:

- Admission webhooks are strictly linked to Kubernetes API Server/etcd **CRUD operations** e.g. Create or Update
  Cluster in etcd.
- Runtime Hooks can be used to define **arbitrary operations**, e.g. Cluster.BeforeUpgrade, Machine.Remediate etc.

In other words, Runtime Hooks are not concerned about “low-level” details of how Kubernetes handles objects in the

> API Server/etcd; Runtime Hooks instead focus on “high-level” events of a Cluster’s lifecycle, or _by opportunity_,

Author: left over, simplified

(e.g by enforcing a maximum timeout for all the Runtime Extensions calls).

### Calling Runtime Extensions

_Note: the code described in this paragraph will be executed only if the Runtime Extension feature flag is set to true._

Cluster API is going to implement calls to registered Runtime Extensions at well-known moments of the Cluster’s lifecycle.

The two key elements that make the implementation of runtime extension calls simple and consistent across
the codebase are:

- The Runtime Hook catalog, providing the info about all the defined Runtime Hooks, supported version and
  corresponding input/output types;
- The Runtime Extensions registry, providing info about the registered Runtime Extensions implementing the
  Runtime Hooks defined above.

Given these two elements, the code for calling Runtime Extensions is:

_Note: this is pseudo code, meant to demonstrate a set of elements described below the example; the exact details
will be defined during implementation, possibly taking advantage of golang generics._

```go
extensions := registry.Get(
    registry.Group("cluster.runtime.cluster.x-k8s.io/v1alpha2”),

>     registry.Hook(&v1alpha2.BeforeUgradeHook{}),

Reviewer 4: Instead of using the "hub" version of the hook as the identifier should we consider using something else instead. I am asking to 

Consider a case where a hook is dropped in a new version but is still present in the old versions  - in such a case using the last version as the "hub" version to identify the list of hooks will not be ideal. Should we explore some other way to maintain a list of all available hook (across all versions). The version specific definitions of the hooks will then be registered against this new identifier.

Author: As written in the note this is pseudo code, so details are up to changes but the TL;DR; is that we should implement the call operation only once vs "duplicating" the code for each new version.
Said that, a possible way forward for the use case when a hook is dropped is to consider the last available version of that hook as "Hub" (there is no single "Hub" version like in APIs, you can choose your preferred version for each hook); other solutions can be figured out during implementation.

Reviewer 3: +1 We should dive into those details during implementation. But I don't think it's a big issue. 

For now it seems to be good enough to just define the latest existing version as the hub for conversion.

In other words, Runtime Hooks are not concerned about “low-level” details of how Kubernetes handles objects in the
API Server/etcd; Runtime Hooks instead focus on “high-level” events of a Cluster’s lifecycle.

Please note that, no matter the similarities in some part of the design, users should not make assumptions about
Runtime Hooks having properties or behaviors typical of Kubernetes admission webhooks unless they are explicitly
defined in the following paragraphs.

#### Runtime SDK rules

As this proposal is based on RESTful APIs, we are using [OpenAPI Specification v3.0.0](https://swagger.io/specification/) [1] 
to document Runtime Hooks supported by Cluster API.

Most specifically, a single OpenAPI document providing specification for all the Runtime Hooks supported by a
Cluster API release will be added to the release artifacts; users can rely on https://editor.swagger.io/ or similar
tools to view the specification, and during implementation we will consider adding similar view to the Cluster API
book as well e.g.

![overview](images/runtime-sdk/swagger-ui.png)

> Each Runtime Hook will be defined by one (or more) RESTful APIs implemented as a `POST` operation; each operation

Reviewer 5: `POST` are generally not idempotent. Was `PUT` or `PATCH` considered?

Reviewer 3: Good question. 

One data point. Kubernetes is using POST too for Validationg/Mutating webhooks.

Reviewer 3: Maybe Kubernetes is using `POST` because for AdmissionRequests they are not creating the same resource all over again. It's a new one every time they call an admission hook.

I think it's similar for us + that we expect when we create multiple "different" resources we are always getting the same response. But we're not create+updating the same resource (only the content is the same)

Author: I will keep this consistent with Kubernetes for now...

additional requests according to the OpenAPI specification for the Runtime Hook you are interested in.

E.g.

```go
// Note: this is pseudo code, meant to demonstrate that implementing a RuntimeExtension requires only minimal scaffolding;
// the exact details will be defined during implementation, possibly taking advantage of Golang generics.

var c = catalog.NewCatalog()

func init() {
    v1alpha2.AddToCatalog(c)
}

func main() {
    ctx := context.Background()

    listener, err := net.Listen("tcp", net.JoinHostPort("127.0.0.1", "8082"))
    if err != nil {
        panic(err)
    }

    // Build an HTTP handler for a given operation, calling a strongly typed func at each request
    beforeUgradeHandler, err := catalogHTTP.NewHandlerBuilder().
        WithCatalog(c).
        AddService(&v1alpha2.DiscoveryHook{}, doDiscovery).
        AddService(&v1alpha2.BeforeUgradeHook{}, doBeforeUpgrade).
        Build()
    if err != nil {
        panic(err)
    }

    srv := &http.Server{
        Handler: BeforeUgradeHandler,
    }
    if err := srv.Serve(listener); err != nil && err != http.ErrServerClosed {
        panic(err)
    }
}

> func doBeforeUpgrade(in *v1alpha2.DiscoveryInput, out *v1alpha2.DiscoveryOutput) error {

Reviewer 3: ```suggestion
func doDiscovery(in *v1alpha2.DiscoveryInput, out *v1alpha2.DiscoveryOutput) error {
```
nit

    out.Items = append(out.Items,
        *v1alpha2.DiscoveryExtension{
            name: "upgradeAddons",
            hook: *v1alpha2.DiscoveryHook{
                apiVersion: "hook.runtime.cluster.x-k8s.io/v1alpha1",
                name: "BeforeUpgrade",
            },

>             timeoutSeconds: 5,

Reviewer 6: Would it be useful to have priority for a hook, for cases where some kind of ordering might be required to co-ordinate between hooks of the same type?

Reviewer 3: In this iteration of the proposal we would like to avoid introducing dependencies like this:

> Instead, unless there's a strong and evident need for it, we are not considering adding support for defining dependencies among Runtime Extensions, being it modeled with something similar to [systemd unit options](https://www.freedesktop.org/software/systemd/man/systemd.unit.html) or alternative approaches.

> The main reason behind that is that such type of feature introduces complexity and creates "pet" like relations across components making the overall system more fragile. This is also consistent with the [avoid dependencies](https://github.com/kubernetes-sigs/cluster-api/blob/9a3a8f557db850b4a67e9e8e0a70cb2d0261caee/docs/proposals/20220221-runtime-SDK.md#avoid-dependencies) recommendation above.

Author: As explained by @sbueringer we are not going to support dependencies between extensions in the first iteration.

  will take care of version conversions, if required.

## Security Model

Following threats were considered:

- Malicious Runtime Extensions being registered 

Mitigation: The same mitigations used for avoiding malicious dynamic webhooks in Kubernetes apply
(defining RBAC rules for the RuntimeExtensionConfiguration assigning this responsibility to cluster admin only).

- Privilege escalation of HTTP Servers running Runtime Extensions

Mitigation: The same mitigations used for any HTTP server deployed in Kubernetes apply
(use distroless base image, do not use privileged pods etc.).

- Tampering of the communication channel between Cluster API controllers and HTTP Servers implementing Runtime Extensions.

Mitigation: The same mitigations used for any HTTP server deployed in Kubernetes apply (use SSL, Network policies etc.).

## Risks and Mitigations

- Building Runtime SDK, Runtime Hooks and Runtime Extensions in sequential steps might lead to reworks.

This is an accepted risk, given the importance of defining a robust SDK before external developers start relying
extensively on this feature.

> ## Alternatives

Reviewer 5: Perhaps, another alternative would be to use a WASM module. A WASM engine could be hosted in process and load a module to handle specific events. There would be no need for grpc based plugins since the WASM module could run sandboxed in-process.

Prior art:
- https://www.openpolicyagent.org/docs/latest/wasm/
- https://istio.io/latest/docs/reference/config/proxy_extensions/wasm-plugin/
- https://docs.kubewarden.io

Author: this alternative is already documented in "Using in-process plugins vs calling external components". I have also experimented on it but then discarded it for the reason described below; Personally, I also find that the golang support for web assembly seems very limiting, and introducing another language/toolchain is not practical.

Reviewer 5: My fault. I missed the WebAssembly call out.

The trade off seems to be operational / perf pain from external dependencies vs. dev / tool chain pain. Personally, the latter would be more interesting to me as that space is quickly improving where the ops / perf of using external services would be forever.

Author: Closing as per March 23th office hours discussion

GA Runtime Hook API versions etc.

**Rule #4: Other than the most recent Runtime Hook versions in each track, older Runtime Hook versions must
be supported after their announced deprecation for a duration of no less than:

- GA: 12 months or 3 releases (whichever is longer)
- Beta: 6 months or 2 releases (whichever is longer)
- Alpha: 0 releases
  **

### Discovery hook

The Discovery hook must be implemented by all the Runtime Extensions servers, and it is responsible to
inform the system about the Runtime Extensions it implements.

When invoked the discovery hook is expected to provide the following answer:

```yaml
status: Success # or Failure
message: "error message if status == failure"
items: # Info about implemented runtime extensions
- name: http-proxy     # Unique name identifying the runtime extension
  hook:
    apiVersion: "hook.runtime.cluster.x-k8s.io/v1alpha1"
    name: "generatePatches"
  timeoutSeconds: 5    # Default value suggested by the RuntimeExtension developers
  failurePolicy: Fail  # Default value suggested by the RuntimeExtension developers 
- ...
```

> Please note the above struct support defining more than one runtime extensions for the same hook, e.g.
> defining more than one "generatePatches" extensions. 

Reviewer 3: ```suggestion
Please note that the above struct supports defining more than one Runtime Extension for the same hook, e.g.
defining more than one "generatePatches" extensions. 
```

nit, (just because the example currently only defines one hook)

---
title: Cluster API Runtime SDK
authors:
  - "@fabriziopandini"
  - "@sbueringer"
reviewers:
  - "@enxebre"
  - "@vincepri"
  - "@CecileRobertMichon"
creation-date: 2022-02-21

> last-updated: 2022-02-21

Reviewer 2: needs to be updated

    * [Runtime SDK rules](#runtime-sdk-rules-1)
    * [Discovery hook](#discovery-hook)
  * [Implementation History](#implementation-history)
  
## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

- **Cluster API Runtime**: identifies the Cluster API execution model, a set of controllers cooperating in managing the 
  workload cluster’s lifecycle.
- **Runtime SDK**: a set of rules, recommendations and fundamental capabilities required to develop Runtime Hooks and 
  Runtime Extensions.
- **Runtime Hook**: a single, well identified, extension point allowing applications built on top of Cluster API to hook
  into specific moments of the workload cluster’s lifecycle, e.g. Cluster.BeforeUpgrade, Machine.BeforeRemediation.
- **Runtime Extension**: an external component which is part of a system/product built on top of Cluster API that can 
  handle requests for a specific Runtime Hook.

## Summary

This proposal introduces the Cluster API Runtime SDK, a set of rules, recommendations and fundamental capabilities
required to implement a new extensibility mechanism that will allow systems, products and services built on top of

> Cluster API to hook into the workload cluster’s lifecycle.

Reviewer 7: ```suggestion
Cluster API to hook into a workload cluster’s lifecycle.
```
